volttron/platform/dbutils/mysqlfuncts.py

```diff
@@ -300,6 +305,103 @@ def query(self, topic_ids, id_name_map, start=None, end=None, skip=0,
                 cursor.close()
         return values
 
+    @contextlib.contextmanager
+    def bulk_insert_data(self):
+        """
+        This function implements the bulk insert requirements for Redshift historian by overriding the
+        DbDriver::bulk_insert_data() in basedb.py and yields necessary data insertion method needed for bulk inserts
+        :yields: insert method
+        """
+        records = []
+
+        def insert_data(ts, topic_id, data):
+            """
+            Inserts data records to the list
+            :param ts: time stamp
+            :type string
+            :param topic_id: topic ID
+            :type string
+            :param data: data value
+            :type any valid JSON serializable value
+            :return: Returns True after insert
+            :rtype: bool
+            """
+#            _log.info("appended record")
+            value = jsonapi.dumps(data)
+            records.append((ts, topic_id, value))
+#            records.append(SQL('({}, {}, {})').format(Literal(ts), Literal(topic_id), Literal(value)))
+            return True
+
+        yield insert_data
+
+        if records:
+            query = f"""
+INSERT INTO {self.data_table} (ts, topic_id, value_string) VALUES(%s, %s, %s)
+ON DUPLICATE KEY UPDATE value_string=VALUES(value_string);
+"""
+            _log.debug(f"calling execute many with records {len(records)}")
+            self.execute_many(query, records)
+
+    @contextlib.contextmanager
+    def bulk_insert_topic(self):
+        """
+        This function implements the bulk insert requirements for Redshift historian by overriding the
+        DbDriver::bulk_insert_topic() in basedb.py and yields necessary data insertion method needed for bulk inserts
+        :yields: insert method
+        """
+        topics = []
+
+        def insert_topic(topic):
+            """
+            Inserts topic records to the list
+            :param topic: topic name
+            :type string
+            :return: Returns topic_id
+            :rtype: int
+            """
+            self.max_topic_id = self.max_topic_id + 1
```

Comment:
We talked about this, however if we have multiple historians pointing at the same database there is a real possibility for over lapping topic id->topic mapping.  This is a use case we didn't think about when we discussed this.

Reply:
Should we make batch insert of topic optional based on configuration?

Reply:
I think batch insert is good

We should do something like this, because of the following conditions

- mysql will autoincrement sequentially (I believe it will lock the table for an insert transaction)

therfore we can 
1. insert batch
2. get last topic entered
3. assign topic_id to the batch based upon order in the list of items

So

batchinsert ([t1, t2, t3])
get_last_id -> 50

then
topics[t3] = 50
topics[t2] = 49
topics[t1] = 48